lib/ipaddr.ml

```diff
@@ -925,18 +933,15 @@ module V6 = struct
       && Domain_name.equal_label labels.(0) "arpa"
       && Domain_name.equal_label labels.(1) "ip6"
     then
-      let b = B128.zero () in
-      let bi = ref 0 in
-      let i = ref 2 in
+      let b = Bytes.create 16 in
       try
-        while !i <= 32 do
-          let x = int_of_char_string labels.(!i) in
-          let y = int_of_char_string labels.(!i + 1) in
-          Bytes.set_uint8 b !bi (Int.logor (Int.shift_left x 4) y);
-          bi := !bi + 1;
-          i := !i + 2
+        for bi = 0 to 15 do
+          let i = 2 * Int.succ bi in
+          let x = int_of_char_string labels.(i) in
+          let y = int_of_char_string labels.(i + 1) in
+          Bytes.set_uint8 b bi (Int.logor (Int.shift_left x 4) y);
         done;
-        Some b
+        Some (S128.of_octets_exn (Bytes.unsafe_to_string b))
```

Comment:
Here we sort of leak the abstraction / have imperative code. Maybe we need a `S128.of_int4` :D